### PR TITLE
Quantum Monte Carlo For Chemistry

### PR DESCRIPTION
**Pull request by yitchen-tim**
*Description of changes:*
Add a new example notebook featuring solving the ground state energy of hydrogen molecule with quantum Monte Carlo. The work is done by WWSO quantum intern Benchen Huang!

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by ajberdy** [comment]
README should be non-empty or deleted. I don't think we need a readme for notebook format

File: examples/hybrid_quantum_algorithms/Quantum_Monte_Carlo_Chemistry/afqmc/classical_afqmc.py [hunk 1]
```diff
@@ -0,0 +1,335 @@
+#!/usr/bin/env
```
Comment (ajberdy) [inline examples/hybrid_quantum_algorithms/Quantum_Monte_Carlo_Chemistry/afqmc/classical_afqmc.py:1]: do we need this here, given that it's a file containing helper functions and is never run directly?

File: examples/hybrid_quantum_algorithms/Quantum_Monte_Carlo_Chemistry/afqmc/classical_afqmc.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+
+    V = np.zeros((num_spin_orbitals, num_spin_orbitals), dtype=np.complex128)
+    for i in range(num_fields):
+        V += np.sqrt(dtau) * xshifted[i] * v_gamma[i]
+    exp_V = expm(V)
+
+    # Note that v_gamma doesn't include the mf_shift, there is an additional term coming from
+    # -(x - xbar)*mf_shift, this term is also a complex value.
+    # cmf = -np.sqrt(dtau)*np.dot(xshifted, mf_shift)
+    # prefactor = np.exp(-dtau*(H_0 - E_0) + cmf)
```
Comment (ajberdy) [inline examples/hybrid_quantum_algorithms/Quantum_Monte_Carlo_Chemistry/afqmc/classical_afqmc.py:187]: are these related to the comments? should be deleted or uncommented?

File: examples/hybrid_quantum_algorithms/Quantum_Monte_Carlo_Chemistry/afqmc/quantum_afqmc_pennylane.py [hunk 3]
```diff
@@ -0,0 +1,704 @@
...
+
+    U_circuit(U)
+    for i in pauli:
+        qml.PauliZ(wires=i)
+
+    qml.adjoint(U_circuit)(U)
+    circuit_second_half_imag(Q, V_T)
+
+
+def pauli_estimate(Q, V_T, U, pauli: list, dev):
```
Comment (ajberdy) [inline examples/hybrid_quantum_algorithms/Quantum_Monte_Carlo_Chemistry/afqmc/quantum_afqmc_pennylane.py:208]: inconsistent type annotations throughout, variables/parameters should generally be snake case. Not sure that the formatting is a priority for helper files though

File: examples/hybrid_quantum_algorithms/Quantum_Monte_Carlo_Chemistry/afqmc/quantum_afqmc_pennylane.py [hunk 4]
```diff
@@ -0,0 +1,704 @@
...
+        pauli_imag(Q, V_T, U, pauli)
+        return qml.probs(range(num_qubits))
+
+    probs_values = compute_imag(Q, V_T, U, pauli)
+    imag = probs_values[0] - probs_values[int(2**num_qubits / 2)]
+
+    return real + 1.0j * imag
+
+
+def qExpect_OneBody(walker, one_bodies, ovlp, V_T, dev):
```
Comment (ajberdy) [inline examples/hybrid_quantum_algorithms/Quantum_Monte_Carlo_Chemistry/afqmc/quantum_afqmc_pennylane.py:244]: non-pythonic casing, same stipulations as above

File: examples/hybrid_quantum_algorithms/Quantum_Monte_Carlo_Chemistry/afqmc/run_q_afqmc.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    Ehf = local_energy(h1e, eri, G, nuclear_repulsion)
+    print(f"The Hartree-Fock energy computed from local_energy is {np.round(Ehf, 10)}.")
+
+    #####################################################################
+    # Execution.                                                        #
+    #####################################################################
+    # Read the hyperparameters
+    hp_file = os.environ["AMZN_BRAKET_HP_FILE"]
+    with open(hp_file) as f:
+        hyperparams = json.load(f)
```
Comment (ajberdy) [inline examples/hybrid_quantum_algorithms/Quantum_Monte_Carlo_Chemistry/afqmc/run_q_afqmc.py:50]: we can replace this with parameters for main with typings and avoid opening the HP_FILE and casting explicitly. I'll add a comment above to demonstrate

File: examples/hybrid_quantum_algorithms/Quantum_Monte_Carlo_Chemistry/afqmc/run_q_afqmc.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+np.set_printoptions(precision=4, edgeitems=10, linewidth=150, suppress=True)
+
+
+from afqmc.classical_afqmc import G_pq, chemistry_preparation, local_energy
+from afqmc.quantum_afqmc_pennylane import qAFQMC
+from braket.jobs import save_job_result
+from braket.jobs.metrics import log_metric
+
+
+def main():
```
Comment (ajberdy) [inline examples/hybrid_quantum_algorithms/Quantum_Monte_Carlo_Chemistry/afqmc/run_q_afqmc.py:18]:
```suggestion
def main(
    num_walkers: int,
    num_steps: int,
    dtau: float,
    max_pool: int,
    q_total_time: str,
):
```
q_total_time will still need to call json.loads

File: examples/hybrid_quantum_algorithms/Quantum_Monte_Carlo_Chemistry/afqmc/quantum_afqmc_pennylane.py (hunk 3 above)
Comment (yitchen-tim) [reply examples/hybrid_quantum_algorithms/Quantum_Monte_Carlo_Chemistry/afqmc/quantum_afqmc_pennylane.py]: @ajberdy In scientifically, it is a convention to use upper and lower cases to differentiate different object in equations. In this example, allowing underscores and upper case adds readability to code because the correspondence to the equations is more clear. I suggest that we can keep the case in variable name.

**Comment by yitchen-tim** [comment]
@mbeach-aws Is it faster to use c5.18xlarge rather than m5.24xlarge? Also, when I run the notebook (with m5.24xlarge NBI) I got 5 min which is different from 2.5 min that you got. Did you use c5.18xlarge for NBI when running the notebook?

Also, the Job takes 102 mins which is a bit too long. As a notebook, should we scale it down to something like 30 min?

**Comment by yitchen-tim** [comment]
I changed the file name from "quantum_afqmc_pennylane" to just "quantum_afqmc" because we only have one implementation.

**Comment by mbeach-aws** [comment]
@yitchen-tim Reason i had was 18xlarge is because I don't have access to 24xlarge in jobs quotas.  Is it in the default job quotas? (c5.18xlarge seems to be)

Yeah the job takes 102min because it does quantum every 5 steps. It could be changed down to say 10 or 20 to speed it up.

**Comment by yitchen-tim** [comment]
@ajberdy Thanks for reviewing! Regarding the case in variable name and type hint:
(1) variable name: In scientific content, it is a convention to use upper and lower cases to differentiate different object in equations. I know code is different from equations, but in this example, allowing underscores and upper case adds readability to code because the correspondence to the equations is more clear. I suggest that we can keep the case in variable name.
(2) type hint: I fixed the inconsistent type hint.

If these two are good, the remaining one thing is just updating the time and cost estimated (we are still working on that)

File: examples/hybrid_quantum_algorithms/Quantum_Monte_Carlo_Chemistry/Quantum_Monte_Carlo_Chemistry.ipynb [hunk 7]
```diff
@@ -0,0 +1,1213 @@
+{
```
Comment (sammcardle30) [inline examples/hybrid_quantum_algorithms/Quantum_Monte_Carlo_Chemistry/Quantum_Monte_Carlo_Chemistry.ipynb:1]: <p>Change 'QAFE' to 'QAEE' x2 in first line of paragraph.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/216/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='30'/>
